core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -310,12 +360,13 @@ private NodesToAllocate buildNodesToAllocate(RoutingAllocation allocation,
 
             Decision decision = forceAllocate ? allocation.deciders().canForceAllocatePrimary(shardRouting, node, allocation) :
                                                 allocation.deciders().canAllocate(shardRouting, node, allocation);
```

Comment:
I believe here you will also want to check out `allocation.debugDecision()` and setting the flag on the allocation to ensure that the deciders are not short-circuited in the case where the `explain` flag is set